### PR TITLE
Update configurability for CA1068

### PR DESCRIPTION
**Pull request by Evangelink**
## Summary

Update documentation for rule CA1068 to mention the 2 new configurability options.

Relates to https://github.com/dotnet/roslyn-analyzers/pull/4474 and https://github.com/dotnet/roslyn-analyzers/pull/4500

### CODE REVIEW

**Comment by Evangelink** [comment]
Tagging @mavasani

**Review by mavasani — APPROVED** [review]

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 1]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
 
 ## Configurability
 
-[Roslyn analyzers issue 2851](https://github.com/dotnet/roslyn-analyzers/issues/2851) tracks making the rule configurable. Once it's implemented, you can configure the API surface on which rule executes to avoid having to use source suppressions.
+This rule has the following configurable options that can be configured for just this rule, for all rules, or for all rules in this category (Security). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:66]:
```suggestion
This rule has the following configurable options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 2]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+You can configure which parts of your codebase to exclude from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.excluded_symbol_names = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace)
+- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format). Each symbol name requires a symbol kind prefix, such as "M:" prefix for methods, `T:` prefix for types, "N:" prefix for namespaces, etc.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:81]:
```suggestion
- Fully qualified names in the symbol's [documentation ID format](../../../csharp/language-reference/language-specification/documentation-comments.md#id-string-format). Each symbol name requires a symbol-kind prefix, such as `M:` for methods, `T:` for types, and `N:` for namespaces.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 3]
```diff
@@ -24,6 +24,8 @@ ms.author: mavasani
 
 A method has a <xref:System.Threading.CancellationToken> parameter that is not the last parameter.
 
+By default, this rule analyzes the entire codebase, but this is [configurable](#excluded-symbols).
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:27]:
```suggestion
By default, this rule analyzes the entire codebase, but this is [configurable](#configurability).
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 4]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+
+- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace)
+- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format). Each symbol name requires a symbol kind prefix, such as "M:" prefix for methods, `T:` prefix for types, "N:" prefix for namespaces, etc.
+- `.ctor` for constructors and `.cctor` for static constructors
+
+Examples:
+
+| Option Value | Summary |
+| --- | --- |
+|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType` | Matches all symbols named 'MyType' in the compilation
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:88]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType` | Matches all symbols named 'MyType'. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 5]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType1|MyType2` | Matches all symbols named either 'MyType1' or 'MyType2' in the compilation
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:89]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType1|MyType2` | Matches all symbols named either 'MyType1' or 'MyType2'. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 6]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS.MyType.MyMethod(ParamType)` | Matches specific method 'MyMethod' with given fully qualified signature
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:90]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS.MyType.MyMethod(ParamType)` | Matches specific method `MyMethod` with the specified fully qualified signature. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 7]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS1.MyType1.MyMethod1(ParamType)|M:NS2.MyType2.MyMethod2(ParamType)` | Matches specific methods 'MyMethod1' and 'MyMethod2' with respective fully qualified signature
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:91]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS1.MyType1.MyMethod1(ParamType)|M:NS2.MyType2.MyMethod2(ParamType)` | Matches specific methods 'MyMethod1' and 'MyMethod2' with the respective fully qualified signatures. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 8]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+
+#### Excluded type names with derived types
+
+You can configure which types, including its derived types, to exclude from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and its derived types, add the following key-value pair to an *.editorconfig* file in your project:
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:95]:
```suggestion
You can configure specific types, including their derived types, to exclude from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and its derived types, add the following key-value pair to an *.editorconfig* file in your project:
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 9]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+
+```ini
+dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Type name only (includes all types with the name, regardless of the containing type or namespace)
+- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format), with an optional `T:` prefix.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:104]:
```suggestion
- Fully qualified names in the symbol's [documentation ID format](../../../csharp/language-reference/language-specification/documentation-comments.md#id-string-format), with an optional `T:` prefix.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 10]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+
+You can configure which types, including its derived types, to exclude from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and its derived types, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Type name only (includes all types with the name, regardless of the containing type or namespace)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:103]:
```suggestion
- Type name only (includes all types with the name, regardless of the containing type or namespace).
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 11]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format), with an optional `T:` prefix.
+
+Examples:
+
+| Option Value | Summary |
+| --- | --- |
+|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType` | Matches all types named 'MyType' and all of its derived types in the compilation
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:110]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType` | Matches all types named 'MyType' and all of their derived types. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 12]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType1|MyType2` | Matches all types named either 'MyType1' or 'MyType2' and all of their derived types in the compilation
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:111]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType1|MyType2` | Matches all types named either 'MyType1' or 'MyType2' and all of their derived types. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 13]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = M:NS.MyType` | Matches specific type 'MyType' with given fully qualified name and all of its derived types
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:112]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = M:NS.MyType` | Matches specific type 'MyType' with given fully qualified name and all of its derived types. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 14]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = M:NS1.MyType1|M:NS2.MyType2` | Matches specific types 'MyType1' and 'MyType2' with respective fully qualified names and all of their derived types
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:113]:
```suggestion
|`dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = M:NS1.MyType1|M:NS2.MyType2` | Matches specific types 'MyType1' and 'MyType2' with the respective fully qualified names, and all of their derived types. |
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 15]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+
+You can configure which parts of your codebase to exclude from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.excluded_symbol_names = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:80]:
```suggestion
- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace).
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 16]
```diff
@@ -61,7 +63,62 @@ If the method is an externally visible public API that is already part of a ship
...
+- Fully qualified names in the symbol's [documentation ID format](https://github.com/dotnet/csharplang/blob/master/spec/documentation-comments.md#id-string-format). Each symbol name requires a symbol kind prefix, such as "M:" prefix for methods, `T:` prefix for types, "N:" prefix for namespaces, etc.
+- `.ctor` for constructors and `.cctor` for static constructors
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:82]:
```suggestion
- `.ctor` for constructors and `.cctor` for static constructors.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md (hunk 3 above)
Comment (Evangelink) [reply docs/fundamentals/code-analysis/quality-rules/ca1068.md:27]: Other configurability might not be related to the configuration of which part of the codebase is analyzed.

Comment (gewarren) [reply docs/fundamentals/code-analysis/quality-rules/ca1068.md]: I wanted to suggest a different arrangement of the configuration section, but I couldn't create a PR to your fork, so I had to commit directly to your branch instead.

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 17]
```diff
@@ -61,7 +63,68 @@ If the method is an externally visible public API that is already part of a ship
...
 ## Configurability
 
-[Roslyn analyzers issue 2851](https://github.com/dotnet/roslyn-analyzers/issues/2851) tracks making the rule configurable. Once it's implemented, you can configure the API surface on which rule executes to avoid having to use source suppressions.
+This rule has the following configurable options that can be configured for just this rule, for all rules, or for all rules in this category (Design). For more information, see [Code quality rule configuration options](../code-quality-rule-options.md).
+
+### Include or exclude selected APIs
+
+Use the following options to configure which parts of your codebase to run this rule on.
+
+- [Include specific API surfaces](#api-surface)
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:72]:
```suggestion
- [Include specific API surfaces](#include-specific-api-surfaces)
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 18]
```diff
@@ -61,7 +63,68 @@ If the method is an externally visible public API that is already part of a ship
...
+You can exclude specific symbols, for example, types methods, from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.excluded_symbol_names = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Symbol name only (includes all symbols with the name, regardless of the containing type or namespace).
+- Fully qualified names in the symbol's [documentation ID format](../../../csharp/language-reference/language-specification/documentation-comments.md#id-string-format). Each symbol name requires a symbol-kind prefix, such as `M:` for methods, `T:` for types, and `N:` for namespaces.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:95]:
```suggestion
- Fully qualified names in the symbol's [documentation ID format](../../../csharp/programming-guide/xmldoc/processing-the-xml-file.md#id-strings). Each symbol name requires a symbol-kind prefix, such as `M:` for methods, `T:` for types, and `N:` for namespaces.
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 19]
```diff
@@ -61,7 +63,68 @@ If the method is an externally visible public API that is already part of a ship
...
+You can configure specific types, including their derived types, to exclude from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and its derived types, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.excluded_type_names_with_derived_types = MyType
+```
+
+Allowed symbol name formats in the option value (separated by `|`):
+
+- Type name only (includes all types with the name, regardless of the containing type or namespace).
+- Fully qualified names in the symbol's [documentation ID format](../../../csharp/language-reference/language-specification/documentation-comments.md#id-string-format), with an optional `T:` prefix.
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:118]:
```suggestion
- Fully qualified names in the symbol's [documentation ID format](../../../csharp/programming-guide/xmldoc/processing-the-xml-file.md#id-strings), with an optional `T:` prefix.
```

**Review by gewarren — COMMENTED** [review]
For build warnings.

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 20]
```diff
@@ -61,7 +63,68 @@ If the method is an externally visible public API that is already part of a ship
...
+
+You can include or exclude APIs from analysis based on their accessibility. For example, to specify that the rule should run only against the non-public API surface, add the following key-value pair to an *.editorconfig* file in your project:
+
+```ini
+dotnet_code_quality.CA1068.api_surface = private, internal
+```
+
+#### Exclude specific symbols
+
+You can exclude specific symbols, for example, types methods, from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an *.editorconfig* file in your project:
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:86]:
```suggestion
You can exclude specific symbols, such as types and methods, from analysis. For example, to specify that the rule should not run on any code within types named `MyType`, add the following key-value pair to an *.editorconfig* file in your project:
```

File: docs/fundamentals/code-analysis/quality-rules/ca1068.md [hunk 21]
```diff
@@ -61,7 +63,68 @@ If the method is an externally visible public API that is already part of a ship
...
+| Option Value | Summary |
+| --- | --- |
+|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType` | Matches all symbols named 'MyType'. |
+|`dotnet_code_quality.CA1068.excluded_symbol_names = MyType1|MyType2` | Matches all symbols named either 'MyType1' or 'MyType2'. |
+|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS.MyType.MyMethod(ParamType)` | Matches specific method `MyMethod` with the specified fully qualified signature. |
+|`dotnet_code_quality.CA1068.excluded_symbol_names = M:NS1.MyType1.MyMethod1(ParamType)|M:NS2.MyType2.MyMethod2(ParamType)` | Matches specific methods 'MyMethod1' and 'MyMethod2' with the respective fully qualified signatures. |
+
+#### Exclude specific types and their derived types
+
+You can configure specific types, including their derived types, to exclude from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and its derived types, add the following key-value pair to an *.editorconfig* file in your project:
```
Comment (gewarren) [inline docs/fundamentals/code-analysis/quality-rules/ca1068.md:109]:
```suggestion
You can exclude specific types and their derived types from analysis. For example, to specify that the rule should not run on any methods within types named `MyType` and their derived types, add the following key-value pair to an *.editorconfig* file in your project:
```

**Review by gewarren — COMMENTED** [review]
A couple more tweaks.

**Pull request merged**